### DISCUSSION TITLE
How to render rich text?

### DISCUSSION THREAD

**Discussion by mrcnw** [strapi/strapi#5876] [Legacy Topics]
Hello,

I have the problem associated with rendering rich text. I am using Gatsby and I have tried react-markdown, but it doesn't work. Only bullet points make let's say `<br />`. How to you deal with WYSIWYG in Strapi?

![image](https://user-images.githubusercontent.com/32802948/79703125-08af9f80-82aa-11ea-9fdf-d12da34759c2.png)

**Comment by DarkTraveler** [comment]
@mrcnw Hey. 
Use this instruction https://strapi.io/blog/how-to-change-the-wysiwyg-in-strapi

**Reply by gonzarascon** [reply]
Hi, I've read this post and it's not clear for me how to change the rich text editor. It appears that the article is incomplete, is there another guide (or repo) that shows how to do this?

**Reply by DarkTraveler** [reply]
@gonzarascon 
Hey. I did on the basis of this article and everything worked out :) What are the problems?

**Reply by gonzarascon** [reply]
I don't really understand what changes should I make to `WysiwygWithErrors/index.js` in the `extensions` folder. Do you have some code I can use as reference?

**Reply by DarkTraveler** [reply]
@gonzarascon hey, 
I opened the instruction and see that for some reason it was cut. I attach files to make it easier for you. I'm sorry, my English is not very.

https://drive.google.com/drive/folders/1Yu8ZOjItA1dg-9G1PMm6oeo7GnPmcId5?usp=sharing

![image](https://user-images.githubusercontent.com/16899697/83333005-be2a3580-a2e1-11ea-96ce-02110f78bf3a.png)

Need both folders from the screenshot

**Reply by DarkTraveler** [reply]
@derrickmehaffy 
Hey. For some reason, the WYSIWYG instruction was trimmed.
https://strapi.io/blog/how-to-change-the-wysiwyg-in-strapi
It used to be more voluminous with code examples. Can anyone update it?

**Reply by DarkTraveler** [reply]
![image](https://user-images.githubusercontent.com/16899697/83333158-98e9f700-a2e2-11ea-8bb2-0b60b86cccdc.png)

**Reply by gonzarascon** [reply]
> @gonzarascon hey,
> I opened the instruction and see that for some reason it was cut. I attach files to make it easier for you. I'm sorry, my English is not very.
> 
> https://drive.google.com/drive/folders/1Yu8ZOjItA1dg-9G1PMm6oeo7GnPmcId5?usp=sharing
> 
> ![image](https://user-images.githubusercontent.com/16899697/83333005-be2a3580-a2e1-11ea-96ce-02110f78bf3a.png)
> 
> Need both folders from the screenshot

Thank you very much for your help! Looking forward to read the full post when the lost text gets back 😊

**Reply by gonzarascon** [reply]
Hey, I found this guide in the Strapi docs: https://strapi.io/documentation/v3.x/guides/registering-a-field-in-admin.html

**Comment by DBosley** [comment]
The rich text that strapi generates by default is just markdown. Use a markdown to html parser to render it on your site.

**Reply by robertpiosik** [reply]
To be more precise. Do it on your frontend application (api consumer) level.

**Comment by derrickmehaffy** [comment]
Hello,

I see you are wanting to ask a question that is not really a bug report or a feature request, questions should be directed to the [discussions](https://github.com/strapi/strapi/discussions)

I have converted this to a discussion for you, we try to keep Github clean and only for bug reports and feature requests.

**Comment by jessecheu** [comment]
I used this resource as a guide for this issue. https://www.newline.co/@dmitryrogozhny/how-to-render-markdown-in-react-with-react-markdown--5d1c3849

**Comment by soupette** [comment]
Hello, you can use markdown-it to render markdown. Here's how we do it in the Strapi admin panel:

1. First install the needed dependencies:

`yarn add markdown-it markdown-it-abbr markdown-it-container markdown-it-deflist markdown-it-emoji markdown-it-footnote markdown-it-ins markdown-it-mark markdown-it-sub markdown-it-sup`

2. Create an `mdRenderer.js` file

```js 
/* eslint-disable prefer-template */
import Markdown from 'markdown-it';
import 'highlight.js/styles/solarized-dark.css';
import hljs from 'highlight.js';
import abbr from 'markdown-it-abbr';
import container from 'markdown-it-container';
import deflist from 'markdown-it-deflist';
import emoji from 'markdown-it-emoji';
import footnote from 'markdown-it-footnote';
import ins from 'markdown-it-ins';
import mark from 'markdown-it-mark';
import sub from 'markdown-it-sub';
import sup from 'markdown-it-sup';

const md = new Markdown({
  html: true, // Enable HTML tags in source
  xhtmlOut: false,
  breaks: false,
  langPrefix: 'language-',
  linkify: true,
  typographer: true,
  // Code from: https://github.com/markdown-it/markdown-it/blob/master/support/demo_template/index.js#L83
  highlight: (str, lang) => {
    if (lang && lang !== 'auto' && hljs.getLanguage(lang)) {
      return (
        '<pre class="hljs language-' +
        md.utils.escapeHtml(lang.toLowerCase()) +
        '"><code>' +
        hljs.highlight(lang, str, true).value +
        '</code></pre>'
      );
    }

    if (lang === 'auto') {
      const result = hljs.highlightAuto(str);

      return (
        '<pre class="hljs language-' +
        md.utils.escapeHtml(result.language) +
        '"><code>' +
        result.value +
        '</code></pre>'
      );
    }

    return '<pre class="hljs"><code>' + md.utils.escapeHtml(str) + '</code></pre>';
  },
})
  .use(abbr)
  .use(container, 'warning')
  .use(container, 'tip')
  .use(deflist)
  .use(emoji)
  .use(footnote)
  .use(ins)
  .use(mark)
  .use(sub)
  .use(sup);

// Code from: https://github.com/markdown-it/markdown-it-footnote/blob/master/index.js#L29
md.renderer.rules.footnote_ref = (tokens, idx, options, env, slf) => {
  const caption = slf.rules.footnote_caption(tokens, idx, options, env, slf);

  return '<sup class="footnote-ref"><span>' + caption + '</span></sup>';
};

md.renderer.rules.footnote_anchor = () => {
  /* ↩ with escape code to prevent display as Apple Emoji on iOS */
  return ' <span class="footnote-backref">\u21a9\uFE0E</span>';
};

export default md;
```

3. Display the markdown:

```js
import React, { memo, useMemo } from 'react';
import md from './mdRenderer';

const PreviewWysiwyg = ({ data }) => {
  const html = useMemo(() => md.render(data || ''), [data]);

  return <div dangerouslySetInnerHTML={{ __html: html }} />;
};

export default memo(PreviewWysiwyg);
```

> In the admin panel we have a Wrapper that provides the needed classnames to display the footnate, tips and warning you can check out the code [here](https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-content-manager/admin/src/components/PreviewWysiwyg/Wrapper.js) 

I hope it helps.

**Reply by LuanColeto** [reply]
Man, this worked perfectly
